tape/src/main/java/com/squareup/tape/QueueElementObserver.java

```diff
@@ -0,0 +1,81 @@
+package com.squareup.tape;
+
+import java.io.IOException;
+
+public interface QueueElementObserver {
+
+  /**
+   * Observe elements flowing through the Tape queue, in order to make handling decisions.
+   * Specifically, this callback will be called anytime a queue element is:
+   * 1) Read from the file (Including during object initialization).
+   * or
+   * 2) Added to the queue.
+   *
+   * See [QueueElementObservation] for the information reported.
+   *
+   * @param observation - Encapsulated data about the element in the queue and the state of the
+   * queue. See [QueueElementObservation].
+   * @throws IOException - If observer decides to throw an exception to stop processing.
```

Comment:
We generally shouldn't use exceptions for control flow. observeElement could return a "KEEP" / "DISCARD" enum value. What's the behavior in the 1) vs 2) case? ie what does discarding does when we're reading from the file? Needs to be documented.

"observe" is probably not the right verb here, as we're not just "listening / observe" but we're also taking an active step and discarding elements.

Reply:
Actually, maybe I mis understood, does "stop processing" mean "ignore that element" or "stop all of the processing" ? What's the scenario?

Reply:
Looks like it's "stop it all, bubble up". I guess exception might make a bit more sense then? heh.